### PR TITLE
🐛 fix: `convert_prefixes` incorrectly merges line breaks

### PR DESCRIPTION
**Pull request by gbaian10**
fix: https://github.com/ljnsn/cz-conventional-gitmoji/issues/208#issuecomment-2408847293

This is a simple test.

```py
convert_prefixes = ["Merge", "Revert", "Squash"]


def convert_prefix(message: str) -> str:
    if any(map(message.startswith, convert_prefixes or [])):
        first_word, *rest = message.split(maxsplit=1)
        if first_word.endswith(":"):
            first_word = first_word[:-1]
        message = f"{first_word.lower()}: {' '.join(rest)}"
        return message
    raise ValueError("No prefix to convert.")


input_message = "Merge: branch 'main' into feature/branch\n\nThis will be merged"
print(convert_prefix(input_message))
```

### CODE REVIEW

File: src/gitmojify/mojify.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def gitmojify(
         The gitmojified message.
     """
     if any(map(message.startswith, convert_prefixes or [])):
-        first_word, *rest = message.split()
+        first_word, *rest = message.split(maxsplit=1)
         if first_word.endswith(":"):
             first_word = first_word[:-1]
         message = f"{first_word.lower()}: {' '.join(rest)}"
```
Comment (ljnsn) [inline src/gitmojify/mojify.py:66]: We don't need to join the rest on space in this case anymore.

```suggestion
        message = f"{first_word.lower()}: {rest}"
```

**Comment by ljnsn** [comment]
Nice, thanks for the PR!

Could you add a test that makes sure this works correctly?

File: src/gitmojify/mojify.py (hunk 1 above)
Comment (gbaian10) [reply src/gitmojify/mojify.py:66]: When using `*rest` to capture other output, the result is a list and it may be empty.
Since the count will only be 0 or 1, instead of using join, it might be written like this?

```py
f"{first_word.lower()}: {rest[0] if rest else ''}"
```

**Comment by gbaian10** [comment]
> Nice, thanks for the PR!
> 
> Could you add a test that makes sure this works correctly?

ok!

File: src/gitmojify/mojify.py (hunk 1 above)
Comment (ljnsn) [reply src/gitmojify/mojify.py:66]: Indeed, you're right. Let's just keep the join in that case, so we don't need a condition. But we can just join on `''` (no space).

**Comment by ljnsn** [comment]
Hm, good question. Since this functionality exists mostly to fix messages auto-generated by git, this shouldn't usually happen, but still good to take into account. We shouldn't modify the message any further, so in this case it should indeed fail.

I think the best would be to first split on `" "` specifically, then check if the first word is in the convert prefixes and only then do the conversion. You can leave that for a separate PR though.

**Comment by gbaian10** [comment]
@ljnsn 
I added tests and merged them into the latest version.
I listed the edge case issue we discussed above in the test's TODO.

**Comment by gbaian10** [comment]
Additionally, I noticed that the version of uv.lock was not upgraded.

image

Is this the responsibility of this hook?

image

---

When using `cz bump`, does the order of the hooks affect this detection? 

For example, if the lock hook check is executed first, it passes because `pyproject.toml` and the lock are synchronized.
However, after `cz bump` upgrades the version and updates `pyproject.toml`, the lock hock is not rechecked.

Can adjusting the order solve this issue, or is it necessary to add the lock hook to pre-push to prevent it?

**Comment by ljnsn** [comment]
Yes indeed, that's something I want to fix as well. It seems that the commitizen github action doesn't run pre-commit hooks (and even if it did, it would fail, since it adjusts files). So I'll replace the action with another step that bumps without pre-commit, runs uv lock, and then commits and pushes. Maybe even make it a separately available github action, since it seems like this would be useful for other using commitizen with uv as well.

**Comment by gbaian10** [comment]
Hmm... that sounds complicated.

I've mostly used Poetry in the past, and it seems like the `poetry.lock` file doesn't include the version number of the project itself, so similar issues haven't occurred.
Recently, many projects have started migrating to UV, so I'm learning it as well.

Please forgive my unfamiliarity with Commitizen.
I was just thinking, does `cz bump` happen during the commit message stage?
If so, is it always impossible to handle it before pre-commit?

~~I think the simplest solution for me is to use pre-push~~

**Comment by gbaian10** [comment]
However, from what you said, it seems like this is a GitHub Action feature, rather than something done locally before pushing.
I'm considering using this feature, but I'm not very familiar with it yet. 😥

**Comment by ljnsn** [comment]
> I've mostly used Poetry in the past, and it seems like the poetry.lock file doesn't include the version number of the project itself, so similar issues haven't occurred.

Yea neither `poetry` nor `pdm` include the version of the project in the lock file. It might also be worth raising this to `uv`, but I have to think about it a bit. Not sure whether the version of the local project is really needed in the lock file.

> I was just thinking, does cz bump happen during the commit message stage?
If so, is it always impossible to handle it before pre-commit?

The bump isn't done via pre-commit, but in a separate CI step using the github action. The uv lock needs to happen after the bump, since otherwise it wouldn't update the version in the lock file. So it cannot be done locally (unless you bump locally of course, but that defeats the purpose of the CI job).

File: tests/gitmojify/test_gitmojify.py [hunk 2]
```diff
@@ -211,6 +211,42 @@ def test_gitmojify_with_convert_prefixes():
...
+    message_in: str, message_out: str, should_raise: bool
+):
+    """Test gitmojify with convert_prefixes and the Merge prefix."""
+    if not should_raise:
+        result = mojify.gitmojify(message_in, convert_prefixes=["Merge"])
+        assert result == message_out
+    else:
+        with pytest.raises(ValueError) as exc_info:
+            mojify.gitmojify(message_in, convert_prefixes=["Merge"])
+        assert str(exc_info.value) == "invalid commit message"
```
Comment (ljnsn) [inline tests/gitmojify/test_gitmojify.py:247]:
```suggestion
        with pytest.raises(ValueError, match="invalid commit message"):
            mojify.gitmojify(message_in, convert_prefixes=["Merge"])
```

**Review by ljnsn — CHANGES_REQUESTED** [review]

File: tests/gitmojify/test_gitmojify.py (hunk 2 above)
Comment (gbaian10) [reply tests/gitmojify/test_gitmojify.py:247]: Does this `match` actually compare whether the error messages are the same, or is it just a message for humans to read? 

I once tried deleting some text within the match to see if it would cause pytest to fail, but the test still passed.

Comment (ljnsn) [reply tests/gitmojify/test_gitmojify.py:247]: It uses `re.search`, so the message doesn't need to match completely, but needs to be contained in the error message.

Comment (gbaian10) [reply tests/gitmojify/test_gitmojify.py:247]: Got it. Since I only tried deleting text and retesting, no wonder the tests always passed. 😅

**Review by ljnsn — APPROVED** [review]

**Comment by gbaian10** [comment]
Not sure if this feature can solve the uv.lock issue.
[post_bump_hooks](https://commitizen-tools.github.io/commitizen/commands/bump/#post_bump_hooks) or pre?

**Comment by ljnsn** [comment]
Ah nice, the pre bump hooks are exactly what we need, since the files are already adjusted but not committed yet.
